In [6]:
import timeit
code="""
from random import randint
array = [randint(1, 100) for _ in range(5000)]
sort_result = merge_sort(array)
result = search_matched_sum(sort_result, 10)

print(result)
"""
setup="""
def search_matched_sum(arr : list, target : int):
    low = 0 
    high = len(arr)-1
    min_index = []
    while (low < high) :
        sum = arr[low]+ arr[high]
        if sum == target:
            min_index = (low,high)
        if target < sum:
            high -= 1
        else: 
            low += 1
    
    return min_index 

def merge_sort(unsorted_array):
    if len(unsorted_array) > 1:
        mid = len(unsorted_array) // 2 
        left = unsorted_array[:mid]
        right = unsorted_array[mid:] 

        merge_sort(left)
        merge_sort(right)

        i = j = k = 0
        
        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                unsorted_array[k] = left[i]
                i += 1
            else:
                unsorted_array[k] = right[j]
                j += 1
            k += 1

        while i < len(left):
            unsorted_array[k] = left[i]
            i += 1
            k += 1

        while j < len(right):
            unsorted_array[k] = right[j]
            j += 1
            k += 1
    return unsorted_array
"""

print(timeit.timeit(stmt = code, setup = setup, number = 500))

(239, 240)
(225, 226)
(254, 255)
(243, 244)
(237, 238)
(261, 262)
(257, 258)
(235, 236)
(255, 256)
(208, 209)
(255, 256)
(231, 232)
(248, 249)
(259, 260)
(249, 250)
(238, 239)
(246, 247)
(232, 233)
(220, 221)
(246, 247)
(282, 283)
(249, 250)
(248, 249)
(256, 257)
(260, 261)
(230, 231)
(244, 245)
(249, 250)
(247, 248)
(267, 268)
(271, 272)
(245, 246)
(264, 265)
(250, 251)
(233, 234)
(243, 244)
(248, 249)
(246, 247)
(270, 271)
(259, 260)
(239, 240)
(253, 254)
(282, 283)
(258, 259)
(257, 258)
(231, 232)
(261, 262)
(265, 266)
(242, 243)
(223, 224)
(263, 264)
(262, 263)
(239, 240)
(234, 235)
(217, 218)
(264, 265)
(256, 257)
(229, 230)
(251, 252)
(263, 264)
(281, 282)
(252, 253)
(256, 257)
(238, 239)
(258, 259)
(252, 253)
(239, 240)
(240, 241)
(245, 246)
(238, 239)
(256, 257)
(268, 269)
(235, 236)
(262, 263)
(247, 248)
(247, 248)
(267, 268)
(239, 240)
(277, 278)
(207, 208)
(222, 223)
(250, 251)
(241, 242)
(254, 255)
(270, 271)
(245, 246)
(226, 227)
(284, 285)
(268, 269)
(255, 256)
(244, 245)

가설 세우기

In [ ]:
#time complexity 

def search_matched_sum(arr : list, target : int):
    low = 0  # 1
    high = len(arr)-1 # 1
    min_index = [] # 1 
    while (low < high) : # n 
        sum = arr[low]+ arr[high] # 1
        if sum == target:  
            min_index = (low,high) # 1
        if target < sum: #logn
            high -= 1 # 1
        else: #logn
            low += 1 # 1
    
    return min_index # 1

# O(nlogn)

def merge_sort(unsorted_array):
    if len(unsorted_array) > 1: # 1 
        mid = len(unsorted_array) // 2 # 1
        left = unsorted_array[:mid] # logn
        right = unsorted_array[mid:] # logn

        merge_sort(left) 
        merge_sort(right)

        i = j = k = 0
        
        while i < len(left) and j < len(right): # n
            if left[i] < right[j]: # 1
                unsorted_array[k] = left[i] # 1
                i += 1 # 1 
            else:
                unsorted_array[k] = right[j] # 1 
                j += 1 # 1
            k += 1 # 1

        while i < len(left): # n 
            unsorted_array[k] = left[i} # 1
            i += 1 # 1
            k += 1 # 1

        while j < len(right): # n
            unsorted_array[k] = right[j] # 1
            j += 1 # 1
            k += 1 # 1
    return unsorted_array

# o(nlogn)

index 값을 찾는 과정에서 순서쌍이 많아질 경우 break는 사용하지 않고, 끝 값을 추가해줬다. 
그 이유는 운 좋게 초반에 target과 일치하여 출력되면 시간을 비교하는데 의미가 없다고 생각했기 때문이다.
그리고, 코드를 보면 좌,우로 값이 나올때까지 low = 0 이 고정이고, high 값만 바뀌며 찾으니 동일하게 움직여야하므로
break 를 통해 동일한 숫자의 처음과 끝을 찾아가는 시간의 차이만 있다고 판단해 사용하지 않았다. 
코드상으로도 비슷하다고 생각했고, 
실제로 큰 차이가 나지 않는 것을 볼 수 있다.

비록 같은 숫자가 많이 나오도록 n이 충분히 커진다면 끝까지 가는데 시간이 오래걸릴 것이다. 
하지만 n을 100만개라 하면 1~100 중 하나의 수는 대략 1만개가 나올 것이다. 컴퓨터의 성능이 1만개를 찾는데는 위 실험결과를 빌리면
number = 10000으로 했을시 대략 17초가 나오고 10만으로 했을시 250초가 나왔다. 따라서 의미있는 차이가 나올려면 n=1천만 이라면 유의미한 결과가 나올 것이다.
number= 500으로 찾았을때 시간이 너무 오래 걸리고 연산에 무리가 오는것으로 판단해, 1개를 찾는 시간으로 비교해 보자면 
nonbreak = 76.30644780000057 
break = 79.05038950000016
으로 의미있는 차이가 나기 시작했다. 

number = 500
tartget = 40

n = 100
0.14208940000003167
0.14378240000019105
0.14378990000000158
0.1333705000001828
0.13109499999995933

n = 500
0.7532057000000805
0.7043235000001005
0.7085231999999451
0.7237426000001506
0.6789393999999902

n = 1000
1.5380316000000676
1.4953877000000375
1.3789401000001362
1.5817486999999346
1.5826501999999891

n = 2000
3.2126837999999225
3.132413900000074
3.2386927000000014
3.2870514000001094
2.9759277000000566

n = 5000
9.016531200000372
7.8749832999997125
8.613183899999967
8.799906799999917
8.804901999999856

n = 10000
18.074544999999944
17.980276299999787
17.980276299999787
18.10916159999988
18.260650799999894

n = 100000
292.7215252999995

break
n = 1000
1.3026923999996143
1.547052199999598
1.3964857000000848
1.4311561000004076
1.4773439999999027

n = 100000
251.5701712


target에 대한 생각도 들어 = 별 차이 없다?

number = 500
target = 10

n= 100
0.12766599999986283
0.11300120000032621
0.1419000999999298
0.15195529999982682
0.14278690000037386

n = 10000
25.334793699999864
25.32763900000009
24.460320899999715
24.54222149999987
23.98898949999966

number = 500
target = 90

n = 100
0.1172286000000895
0.13807859999997163
0.12528689999999187
0.12410910000016884
0.14094280000017534

n = 10000
24.520433200000298
24.688933600000382
24.836597800000163
24.354138800000328
24.372424000000137

In [5]:
import timeit
code="""
from random import randint
array = [randint(1, 100) for _ in range(100)]
sort_result = ordered_search_matched_sum(array)
result = search_matched_sum(sort_result, 90)
print(result)
"""
setup="""
def search_matched_sum(arr : list, target : int):
  min_index = ()
  for i in range(0,len(arr)):
    for j in range(i+1,len(arr)):
      sum = arr[i] + arr[j]
      if sum == target:
        min_index = (i,j)
  return min_index      

def ordered_search_matched_sum(arr : list):
    
  for i in range(len(arr)):
      min_idx = i
      for j in range(i+1, len(arr)):
          if arr[min_idx] > arr[j]:
              min_idx = j

      arr[i], arr[min_idx] = arr[min_idx], arr[i]
  return arr   
"""

print(timeit.timeit(stmt = code, setup = setup, number = 500))


(982, 984)
(980, 995)
(986, 988)
(978, 985)
(984, 987)
(969, 995)
(957, 986)
(963, 992)
(982, 988)
(979, 985)
(962, 973)
(996, 998)
(975, 999)
(971, 994)
(967, 993)
(985, 993)
(965, 983)
(969, 979)
(990, 992)
(949, 959)
(967, 988)
(985, 993)
(986, 989)
(970, 986)
(997, 999)
(983, 999)
(968, 975)
(993, 998)
(989, 991)
(973, 984)
(982, 996)
(977, 990)
(968, 992)
(978, 985)
(982, 992)
(995, 997)
(970, 993)
(964, 998)
(986, 988)
(984, 988)
(993, 999)
(988, 992)
(958, 982)
(972, 984)
(972, 999)
(984, 999)
(977, 989)
(978, 997)
(968, 981)
(983, 999)
(986, 999)
(980, 991)
(991, 997)
(988, 997)
(964, 970)
(994, 997)
(984, 985)
(980, 989)
(973, 988)
(977, 999)
(986, 987)
(971, 978)
(982, 993)
(986, 998)
(953, 969)
(939, 943)
(974, 979)
(960, 986)
(972, 993)
(992, 999)
(967, 980)
(966, 970)
(980, 994)
(979, 996)
(985, 998)
(996, 999)
(989, 990)
(976, 978)
(976, 996)
(972, 992)
(978, 999)
(980, 997)
(983, 995)
(979, 994)
(975, 999)
(995, 996)
(986, 999)
(952, 958)
(969, 971)
(981, 993)
(978, 985)

In [ ]:
#time complexity

def search_matched_sum(arr : list, target : int):
  min_index = () # 1
  for i in range(0,len(arr)): # n
    for j in range(i+1,len(arr)): # n
      sum = arr[i] + arr[j] # 1
      if sum == target: # 1
        min_index = (i,j) # 1 
  return min_index      

# O(n^2)

def ordered_search_matched_sum(arr : list): 
    
  for i in range(len(arr)): # n
      min_idx = i # 1
      for j in range(i+1, len(arr)): # n 
          if arr[min_idx] > arr[j]: # 1
              min_idx = j # 1

      arr[i], arr[min_idx] = arr[min_idx], arr[i] # 1
  return arr   

# O(n^2)

number = 500
target = 40

n = 100
0.46710129999974015
0.46618909999961033
0.47381029999996827
0.46779179999975895
0.4557331999994858

n = 500 
10.350092299999233
10.427435599999626
10.505270399999063
10.424789499998951
10.386153900000863

n = 1000
41.65346390000013
41.98349629999939
42.29615729999932
42.11654650000128
42.30129119999947

n = 10000
164.5776079999996
164.72713440000007
166.8860007000003
164.35607189999973
164.09294890000092

break
n = 1000

36.630171799999516
36.182942899999034
36.42944560000251
35.90659140000207
35.983675600000424

n = 10000
1759.519408199998

number = 500
target = 10

n = 100
0.15128879999974743
0.14693459999762126
0.14611049999803072
0.14400899999964167
0.14936630000011064

n = 1000
13.973208899999008
13.919022900001437
13.986573599999247
14.022261599999183
14.078137300002709

target 차이 별로 없다.

number = 500
target = 90

n = 100
0.15957049999997253
0.14935609999884036
0.14872229999673436
0.1466206000004604
0.14780429999882472

n = 1000
14.341833300000872
13.99429130000135
14.156960600001185
14.19669520000025
14.3004331000011